File: java-common/src/net/devrieze/util/CachingHandleMap.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package net.devrieze.util;

import net.devrieze.util.db.DBHandleMap;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Closeable;
import java.io.IOException;
import java.lang.ref.WeakReference;
import java.sql.SQLException;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.Map.Entry;
import java.util.concurrent.CopyOnWriteArraySet;


/**
 * A {@link DBHandleMap} that uses {@link WeakReference WeakReferences} to store
 * its results. Note that the {@link WrappingIterator#remove() remove} method on
 * the iterator is not efficient if the elements do not implement Handle as it
 * will loop through the cache to find the underlying value.
 *
 * @author Paul de Vrieze
 * @param <V> The type of the elements in the map.
 */
public final class CachingHandleMap<V, T extends Transaction> extends AbstractTransactionedHandleMap<V, T> implements Closeable, AutoCloseable {


  private class WrappingIterator implements AutoCloseableIterator<V> {

    private final Iterator<V> mIterator;
    private V mLast;

    public WrappingIterator(Iterator<V> pIterator) {
      mIterator = pIterator;
    }

    @Override
    public boolean hasNext() {
      return mIterator.hasNext();
    }

    @Override
    public V next() {
      V result = mIterator.next();
      putCache(result);
      mLast = result;
      return result;
    }

    @Override
    public void remove() {
      if (mLast!=null) {
        synchronized(mCacheHandles) {
          if (mLast instanceof Handle<?>) {
            invalidateCache(((Handle<? extends V>)mLast));
          } else {
            for (int i = 0; i < mCacheHandles.length; i++) {
              V value = mCacheValues[i];
              if(value!=null && value.equals(mLast)) {
                mCacheHandles[i]=-1;
                mCacheValues[i]=null;
                break;
              }
            }
          }
        }
      }
      mIterator.remove();
    }

    @Override
    public void close() throws Exception {
      if (mIterator instanceof AutoCloseable) {
        ((AutoCloseable) mIterator).close();
      }
    }

  }

  private class WrappingIterable implements Iterable<V> {

    private final Iterable<V> delegateIterable;

    @Override
    public Iterator<V> iterator() {
      return new WrappingIterator(delegateIterable.iterator());
    }

    public WrappingIterable(final Iterable<V> iterable) {
      delegateIterable = iterable;
    }
  }

  int mCacheHead = 0;
  final long[] mCacheHandles;
  final V[] mCacheValues;
  final CopyOnWriteArraySet<Handle<? extends V>> mPendingHandles = new CopyOnWriteArraySet<>();

  final TransactionedHandleMap<V,T> mDelegate;

  public CachingHandleMap(TransactionedHandleMap<V, T> delegate, final int cacheSize) {
    mDelegate = delegate;
    mCacheHandles = new long[cacheSize];
    mCacheValues = (V[]) new Object[cacheSize];
  }

  @Override
  public T newTransaction() {
    return mDelegate.newTransaction();
  }

  @Override
  public Handle<V> put(T transaction, V value) throws SQLException {
    Handle<V> handle = mDelegate.put(transaction, value);
    putCache(handle, value);
    return handle;
  }

  private void putCache(V pValue) {
    if (pValue instanceof HandleAware) {
      putCache(((HandleAware<V>)pValue), pValue);
    }
  }

  private void putCache(Handle<? extends V> pHandle, V pValue) {
    if (pValue!=null) { // never store null
      synchronized (mCacheHandles) {
        final int pos = mCacheHead;
        final long handle = pHandle.getHandle();
        if (mCacheHandles[pos] != handle) {
          removeFromCache(handle);
        }
        mCacheHandles[pos] = handle;
        mCacheValues[pos] = pValue;
        mCacheHead = (pos + 1) % mCacheValues.length;
      }
    }
  }

  public V getFromDelegate(T pTransaction, Handle<? extends V> pHandle) throws SQLException {
    mPendingHandles.add(pHandle); // internal locking so no locking needed here
    try {
      return mDelegate.get(pTransaction, pHandle);
    } finally {
      mPendingHandles.remove(pHandle);
    }
  }

  public V getFromDelegate(Handle<? extends V> pHandle) throws SQLException {
    mPendingHandles.add(pHandle); // internal locking so no locking needed here
    try {
      return mDelegate.get(pHandle);
    } finally {
      mPendingHandles.remove(pHandle);
    }
  }

  private V getFromCache(long handle) {
    synchronized (mCacheHandles) {
      for (int i = 0; i < mCacheHandles.length; i++) {
        if (mCacheHandles[i] == handle) {
          return mCacheValues[i];
        }
      }
      return null;
    }
  }

  @Override
  public V get(long pHandle) {
    ComparableHandle<V> handle = Handles.<V>handle(pHandle);
    if (isPending(handle)) { // don't cache pending values
      return mDelegate.get(handle);
    }
    long key = pHandle;
    final V val;
    synchronized(mCacheHandles) {
      val = getFromCache(pHandle);

      if (val!=null) {
        return val;
      }
      try {
        return storeInCache(handle, getFromDelegate(handle));
      } catch (SQLException e) {
        throw new RuntimeException(e);
      }
    }
  }

  @Override
  @Nullable public V get(@NotNull T transaction, Handle<? extends V> handle) throws SQLException {
    Long key = Long.valueOf(handle.getHandle());
    final V val;
    synchronized(mCacheHandles) {
      val = getFromCache(handle.getHandle());
      if (val!=null) {
        return val;
      }
      return storeInCache(Handles.handle(handle), mDelegate.get(transaction, handle));
    }
  }

  /**
   * @deprecated Don't use this, It has no transaction support and doesn't work with large element counts.
   */
  @NotNull
  @Deprecated
  @Override
  public <T> T[] toArray(final T[] a) {
    return mDelegate.toArray(a);
  }

  @Override
  public boolean contains(final T transaction, final Handle<? extends V> handle) throws SQLException {
    if (getFromCache(handle.getHandle())!=null) {
      return true;
    }
    return mDelegate.contains(transaction, handle);
  }

  @Override
  public boolean contains(final T transaction, final Object obj) throws SQLException {
    if (obj instanceof Handle) {
      return contains(transaction, (Handle<? extends V>) obj);
    }
    return mDelegate.contains(transaction, obj);
  }

  @Override
  public int size() {
    return mDelegate.size();
  }

  @Override
  public boolean isEmpty() {
    return mDelegate.isEmpty();
  }

  @Override
  public V set(T transaction, Handle<? extends V> handle, V value) throws SQLException {
    invalidateCache(handle);
    return storeInCache(handle, mDelegate.set(transaction, handle, value));
  }

  private V storeInCache(Handle<? extends V> pHandle, V pV) {
    if (! isPending(pHandle)) {
      final long handle = pHandle.getHandle();
      synchronized (mCacheHandles) {
        removeFromCache(handle); // remove whatever old value was there
        putCache(pHandle, pV);
      }
    }
    return pV;
  }

  private void removeFromCache(final long handle) {
    boolean doAssert = false;
    assert doAssert=true;
    for (int i = 0; i < mCacheHandles.length; i++) {
      if (mCacheHandles[i]==handle) {
        mCacheHandles[i]=-1;
        mCacheValues[i]=null;
        if (!doAssert) {
          return;
        }
      }
    }
  }

  private boolean isPending(final Handle<? extends V> pHandle) {
    return mPendingHandles.contains(pHandle);
  }

  @Override
  public void invalidateCache(final Handle<? extends V> pHandle) {
    long handle = pHandle.getHandle();
    removeFromCache(handle);
  }

  @Override
  public void invalidateCache() {
    synchronized (mCacheHandles) {
      Arrays.fill(mCacheHandles,-1);
      Arrays.fill(mCacheValues, null);
      mCacheHead=0;
    }
  }

  @Deprecated
  public V getUncached(T pTransaction, ComparableHandle<V> pHandle) throws SQLException {
    return storeInCache(pHandle, mDelegate.get(pTransaction, pHandle));
  }

  @Override
  public boolean remove(T transaction, Handle<? extends V> handle) throws SQLException {
    removeFromCache(handle.getHandle());
    return mDelegate.remove(transaction, handle);
  }

  @Deprecated
  @Override
  public boolean remove(final Object o) {
    if (o instanceof Handle) {
      return remove((Handle<? extends V>)o);
    }
    invalidateCache();
    return mDelegate.remove(o);
  }

  @Deprecated
  @Override
  public boolean retainAll(final Collection<?> c) {
    invalidateCache();
    return mDelegate.retainAll(c);
  }

  @Override
  public void clear(final T transaction) throws SQLException {
    invalidateCache();
    mDelegate.clear(transaction);
  }

  @Override
  public AutoCloseableIterator<V> iterator(T transaction, boolean readOnly) {
    return new WrappingIterator(mDelegate.iterator(transaction, readOnly));
  }

  @Override
  public Iterable<V> iterable(final T transaction) {
    return new WrappingIterable(mDelegate.iterable(transaction));
  }

  @Override
  public void close() throws IOException {
    invalidateCache();
    if (mDelegate instanceof Closeable) {
      ((Closeable) mDelegate).close();
    } else if (mDelegate instanceof AutoCloseable) {
      try {
        ((AutoCloseable) mDelegate).close();
      } catch (Exception e) {
        throw new RuntimeException(e);
      }
    }
  }

}

File: java-common/src/main/kotlin/net/devrieze/util/CachingHandleMap.kt
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package net.devrieze.util

import java.io.Closeable
import java.io.IOException
import java.sql.SQLException
import java.util.*
import java.util.concurrent.CopyOnWriteArraySet


/**
 * A [OldDBHandleMap] that uses [WeakReferences][WeakReference] to store
 * its results. Note that the [remove][WrappingIterator.remove] method on
 * the iterator is not efficient if the elements do not implement Handle as it
 * will loop through the cache to find the underlying value.

 * @author Paul de Vrieze
 * *
 * @param  The type of the elements in the map.
 */
open class CachingHandleMap<V:Any, T : Transaction>(
      protected open val delegate: MutableTransactionedHandleMap<V, T>,
      cacheSize: Int,
      val handleAssigner: (V, Handle<V>)->V) : AbstractTransactionedHandleMap<V, T>(), Closeable, AutoCloseable {

  constructor(delegate: MutableTransactionedHandleMap<V, T>,
              cacheSize: Int): this(delegate, cacheSize, { v, h -> HANDLE_AWARE_ASSIGNER(v,h) })

  private open inner class WrappingIterator(private val transaction:T, protected open val iterator: Iterator<V>) : MutableAutoCloseableIterator<V> {
    private var last: V? = null

    override final fun hasNext(): Boolean {
      return iterator.hasNext()
    }

    override final fun next(): V {
      val result = iterator.next()
      putCache(transaction, result)
      last = result
      return result
    }

    @Throws(IOException::class)
    override fun close() {
      val it = iterator
      if (it is AutoCloseable) {
        try {
          it.close()
        } catch (e: IOException) {
          throw e
        } catch (e: Exception) {
          throw RuntimeException(e)
        }

      }
    }

    override fun remove() {
      throw UnsupportedOperationException()
    }
  }

  private inner class WrappingMutableIterator(transaction:T, iterator: MutableIterator<V>) : WrappingIterator(transaction, iterator), MutableIterator<V>, AutoCloseableIterator<V> {
    override protected val iterator:MutableIterator<V> get() = super.iterator as MutableIterator<V>
    private var last: V? = null

    override fun remove() {
      if (last != null) {
        synchronized (cacheHandles) {
          val last = last
          if (last is Handle<*>) {
            invalidateCache(last as Handle<V>)
          } else {
            for (i in cacheHandles.indices) {
              val value = cacheValues[i]
              if (value != null && value == last) {
                cacheHandles[i] = -1
                cacheValues[i] = null
                break
              }
            }
          }
        }
      }
      iterator.remove()
    }

    @Throws(IOException::class)
    override fun close() {
      val it = iterator
      if (it is AutoCloseable) {
        try {
          it.close()
        } catch (e: IOException) {
          throw e
        } catch (e: Exception) {
          throw RuntimeException(e)
        }

      }
    }

  }

  private inner class WrappingIterable(private val transaction:T, private val delegateIterable: Iterable<V>) : MutableIterable<V> {

    override fun iterator(): MutableIterator<V> {
      return WrappingIterator(transaction, delegateIterable.iterator())
    }
  }

  internal var cacheHead = 0
  internal val cacheHandles: LongArray
  internal val cacheValues: Array<V?>
  internal val pendingHandles = CopyOnWriteArraySet<Handle<V>>()

  init {
    cacheHandles = LongArray(cacheSize)
    @Suppress("UNCHECKED_CAST")
    cacheValues = arrayOfNulls<Any>(cacheSize) as Array<V?>
  }

  @Throws(SQLException::class)
  override fun <W : V> put(transaction: T, value: W): ComparableHandle<W> {
    val handle = delegate.put(transaction, value)
    putCache(transaction, handle, value)
    return handle
  }

  @Suppress("UNCHECKED_CAST")
  protected fun putCache(transaction: T, pValue: V) {
    if (pValue is Handle<*>) {
      putCache(transaction, pValue as Handle<V>, pValue)
    } else if (pValue is ReadableHandleAware<*>) {
      putCache(transaction, (pValue as ReadableHandleAware<V>).getHandle(), pValue)
    }
  }

  private fun putCache(transaction: T, handle: Handle<out V>, nonUpdatedValue: V?) {
    if (nonUpdatedValue != null) { // never store null
      if (handle.valid) {
        val updatedValue = handleAssigner(nonUpdatedValue, handle)

        // Don't cache if the updated value has a handle that does not match
        if (updatedValue is ReadableHandleAware<*> && updatedValue.getHandle()!=handle) return

        synchronized (cacheHandles) {
          transaction.addRollbackHandler({ invalidateCache(handle) })
          val pos = cacheHead
          val handle = handle.handleValue
          if (cacheHandles[pos] != handle) {
            removeFromCache(handle)
          }
          cacheHandles[pos] = handle
          cacheValues[pos] = updatedValue
          cacheHead = (pos + 1) % cacheValues.size

        }
      }
    }
  }

  @Throws(SQLException::class)
  fun getFromDelegate(transaction: T, pHandle: Handle<V>): V? {
    pendingHandles.add(pHandle) // internal locking so no locking needed here
    try {
      return delegate[transaction, pHandle]
    } finally {
      pendingHandles.remove(pHandle)
    }
  }

  private fun getFromCache(handle: Long): V? {
    synchronized (cacheHandles) {
      for (i in cacheHandles.indices) {
        if (cacheHandles[i] == handle) {
          return cacheValues[i]
        }
      }
      return null
    }
  }

  @Throws(SQLException::class)
  override fun get(transaction: T, handle: Handle<out V>): V? {
    var value: V?
    synchronized (cacheHandles) {
      value = getFromCache(handle.handleValue)
      if (value != null) {
        return value
      }
      value = delegate[transaction, handle]
      return value?.apply { storeInCache(transaction, Handles.handle(handle), this) }
    }
  }

  @Throws(SQLException::class)
  override fun contains(transaction: T, handle: Handle<out V>): Boolean {
    if (getFromCache(handle.handleValue) != null) {
      return true
    }
    return delegate.contains(transaction, handle)
  }

  @Throws(SQLException::class)
  override fun containsElement(transaction: T, element: Any): Boolean {
    if (element is Handle<*>) {
      @Suppress("UNCHECKED_CAST")
      return contains(transaction, element as Handle<V>)
    }
    return delegate.containsElement(transaction, element)
  }

  @Throws(SQLException::class)
  override fun set(transaction: T, handle: Handle<out V>, value: V): V? {
    invalidateCache(handle)
    delegate.set(transaction, handle, value)
    return storeInCache(transaction, handle, value)
  }

  private fun storeInCache(transaction: T, pHandle: Handle<out V>, pV: V): V {
    if (!isPending(pHandle)) {
      val handle = pHandle.handleValue
      synchronized (cacheHandles) {
        removeFromCache(handle) // remove whatever old value was there
        putCache(transaction, pHandle, pV)
      }
    }
    return pV
  }

  private fun removeFromCache(handle: Long) {
    for (i in cacheHandles.indices) {
      if (cacheHandles[i] == handle) {
        cacheHandles[i] = -1
        cacheValues[i] = null
        assert(getFromCache(handle) == null)
        return
      }
    }
  }

  private fun isPending(handle: Handle<out V>): Boolean {
    return pendingHandles.contains(handle)
  }

  override fun invalidateCache(handle: Handle<out V>) {
    removeFromCache(handle.handleValue)
  }

  override fun invalidateCache() {
    synchronized (cacheHandles) {
      Arrays.fill(cacheHandles, -1)
      Arrays.fill(cacheValues, null)
      cacheHead = 0
    }
  }

  @Deprecated("")
  @Throws(SQLException::class)
  fun getUncached(transaction: T, pHandle: ComparableHandle<V>): V? {
    return delegate[transaction, pHandle].apply {
      if (this!=null)
        storeInCache(transaction, pHandle, this)
      else {
        removeFromCache(pHandle.handleValue)
      }
    }
  }

  @Throws(SQLException::class)
  override fun remove(transaction: T, handle: Handle<out V>): Boolean {
    removeFromCache(handle.handleValue)
    return delegate.remove(transaction, handle)
  }

  @Throws(SQLException::class)
  override fun clear(transaction: T) {
    invalidateCache()
    delegate.clear(transaction)
  }

  override fun iterator(transaction: T, readOnly: Boolean): MutableAutoCloseableIterator<V> {
    return WrappingIterator(transaction, delegate.iterator(transaction, readOnly))
  }

  override fun iterable(transaction: T): MutableIterable<V> {
    return WrappingIterable(transaction, delegate.iterable(transaction))
  }

  @Throws(IOException::class)
  override fun close() {
    invalidateCache()
    if (delegate is Closeable) {
      (delegate as Closeable).close()
    } else if (delegate is AutoCloseable) {
      try {
        (delegate as AutoCloseable).close()
      } catch (e: Exception) {
        throw RuntimeException(e)
      }

    }
  }
}
